edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/metrics/AvaliabilityMetrics.cs

```diff
@@ -6,28 +6,33 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Core.Metrics
     using System.Collections.Generic;
     using System.IO;
     using System.Linq;
-    using System.Runtime.CompilerServices;
+    using System.Threading.Tasks;
     using Microsoft.Azure.Devices.Edge.Util;
     using Microsoft.Azure.Devices.Edge.Util.Metrics;
     using Microsoft.Extensions.Logging;
 
-    public interface IAvailabilityMetric
-    {
-        void ComputeAvailability(ModuleSet desired, ModuleSet current);
-    }
-
-    public class AvailabilityMetrics : IAvailabilityMetric
+    public class AvailabilityMetrics : IAvailabilityMetric, IDisposable
     {
         readonly IMetricsGauge running;
         readonly IMetricsGauge expectedRunning;
-        readonly ISystemTime time;
-        readonly ILogger log = Logger.Factory.CreateLogger<Availability>();
+        readonly ISystemTime systemTime;
+        readonly ILogger log = Logger.Factory.CreateLogger<AvailabilityMetrics>();
+
+        // This allows edgeAgent to track its own avaliability. If edgeAgent shutsdown unexpectedly, it can look at the last checkpoint time to determine its previous avaliability.
+        readonly TimeSpan checkpointFrequency = TimeSpan.FromMinutes(5);
+        readonly PeriodicTask updateCheckpointFile;
+        readonly string checkpointFile = "avaliability_checkpoint";
```

Comment:
Where does this file get created? In the current working directory? We should make this configurable if possible.